v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,240 @@
+	uploadPath           string         // uploadPath is the configured HTTP resource path for upload
+	uploadSizeLimitBytes int64          // uploadSizeLimitBytes will enforce a maximum payload size to be uploaded
+}
+
+func newEventBuffer(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) *eventBuffer {
+	return &eventBuffer{
+		buffer:               make(chan []byte, bufferSizeLimitEvents),
+		payload:              newPayloadBuffer(),
+		client:               client,
+		uploadPath:           uploadPath,
+		uploadSizeLimitBytes: uploadSizeLimitBytes,
+	}
+}
+
+// Reconfigure updates the user configurable values
+func (b *eventBuffer) Reconfigure(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	b.client = client
+	b.uploadPath = uploadPath
+	b.uploadSizeLimitBytes = uploadSizeLimitBytes
+
+	if int64(cap(b.buffer)) == bufferSizeLimitEvents {
+		return
+	}
+
+	close(b.buffer)
+	newBuffer := make(chan []byte, bufferSizeLimitEvents)
+
+	for event := range b.buffer {
+		if int64(len(event)) < b.uploadSizeLimitBytes {
+			push(newBuffer, event)
+			continue
+		}
+
+		var decoded EventV1
+		if err := json.Unmarshal(event, &decoded); err != nil {
+			continue
+		}
+		if err := b.dropNDCache(&decoded, event); err != nil {
+			if !errors.As(err, &droppedNDCache{}) {
+				continue
+			}
+			event, err = json.Marshal(decoded)
+			if err != nil {
+				continue
+			}
+		}
+		push(newBuffer, event)
+	}
+	b.buffer = newBuffer
+}
+
+// Push attempts to add a new event to the buffer
+func (b *eventBuffer) Push(event EventV1) error {
+	encoded, err := json.Marshal(event)
+	if err != nil {
+		return err
+	}
+
+	err = b.dropNDCache(&event, encoded)
+	if err != nil {
+		if !errors.As(err, &droppedNDCache{}) {
+			return err
+		}
+	}
+
+	// This only blocks when the buffer is being reconfigured
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	push(b.buffer, encoded)
+
+	return err
+}
+
+// dropNDCache checks if an event is bigger than UploadSizeLimitBytes and
+// tries to drop the NDBuiltinCache. If the event is still too big, drop the event.
+func (b *eventBuffer) dropNDCache(event *EventV1, encoded []byte) error {
+	if int64(len(encoded)) < b.uploadSizeLimitBytes {
+		return nil
+	}
+
+	if event.NDBuiltinCache == nil {
+		return fmt.Errorf("upload chunk size (%d) exceeds upload_size_limit_bytes (%d)",
+			int64(len(encoded)), b.uploadSizeLimitBytes)
+	}
+	event.NDBuiltinCache = nil
+	encoded, err := json.Marshal(event)
+	if err != nil {
+		return err
+	}
+	if int64(len(encoded)) > b.uploadSizeLimitBytes {
+		return fmt.Errorf("upload chunk size (%d) exceeds upload_size_limit_bytes (%d)",
+			int64(len(encoded)), b.uploadSizeLimitBytes)
+	}
+
+	return droppedNDCache{}
+}
+
+func push[T any](ch chan T, data T) {
+	select {
+	case ch <- data:
+	default:
+		<-ch
+		ch <- data
+	}
```

Comment:
If we cannot write to the channel, because its buffer size is exceeded, we'll drop an event (<-ch) and write our event instead (ch <- data). Am I reading this right?

Reply:
You are correct :) I added a comment to help clarify it as well.

Reply:
Thanks. Do you think dropping events is common or uncommon? I could imagine that this is something to log, and even to mention in the README? Like, it seems like a Con to me that events would get dropped when the buffer is full.

In theory, I could imagine different behaviours when the buffer is full -- we could stop serving requests until the buffer has been flushed successfully. Now, I don't mean that you should change the logic to that, but it's perhaps something to call out loudly in docs, "When the buffer is full, oldest events will be dropped".

Reply:
Not sure how common this would be, I think that depends on the configured event limit and how many events come in. I'd assume this could get noisy, looks like the current buffer has a metric it increments for a similar scenario called `logBufferSizeLimitExDropCounterName`. I updated the PR adding a new metric `logBufferEventLimitExDropCounterName` that gets incremented if an event gets dropped. What do you think of that?

Updated the README.md to mention dropping events, it is also mentioned in the configuration docs with the line "OPA will drop old events from the log if this limit is exceeded." by the new `buffer_size_limit_events` option which can never be zero. So hopefully that is direct enough that this buffer will drop events at some point.

Reply:
> I updated the PR adding a new metric logBufferEventLimitExDropCounterName that gets incremented if an event gets dropped. What do you think of that?

Sounds like a good compromise! Not too noisy, but discoverable, not silent. :+1: